### DISCUSSION TITLE
how to insert a character in normal mode

### DISCUSSION THREAD

**Discussion by satorkain** [helix-editor/helix#8481] [Q&A] [answered]
Let's say I want to insert a semicolon at the end of the line when I press Enter (in normal mode). I could not find in the documentation how to insert any character.

`ret = ["goto_line_end", "insert_mode", ";", "normal_mode"]`

**Comment by kirawi** [comment]
#4709

**Reply by satorkain** [reply]
My cfg:
`ret = ["goto_line_end", "insert_mode", "@;", "normal_mode"]
` 

and I'm getting an error: 
`No command named '@;'`

**Reply by kirawi** [reply]
> My cfg:
> ret = ["goto_line_end", "insert_mode", "@;", "normal_mode"] 
> 
> and I'm getting an error:
> No command named '@;'

It's not implemented yet.

**Comment by qingshan** [comment, marked as answer]
you could use `:pipe` / `:insert-output` / `:append-output` commands to change / insert / append text:

```toml
ret = ["insert_at_line_end", ":insert-output printf ';'", "normal_mode"]
```

**Reply by satorkain** [reply]
It helped. For Windows, the line looks like this:
`ret = ["insert_at_line_end", ":insert-output echo ;", "normal_mode", "collapse_selection", "delete_selection"]`

However, this adds a line break at the end, which I fixed by deleting it. `"collapse_selection", "delete_selection".`

**Reply by satorkain** [reply]
It works correctly this way without adding a new line
`ret = ["insert_at_line_end", ":pipe echo ;", "normal_mode", "collapse_selection"]`
